### PR TITLE
Prioritize print-color-mode-supported over PWG when creating color mo…

### PR DESCRIPTION
**Pull request by zdohnal**
…dels

See https://github.com/apple/cups/pull/5722

### CODE REVIEW

**Review by michaelrsweet — COMMENTED** [review]
Not sure about the reasons for this change, as print-color-mode doesn't actually convey color space support for a printer, just the basic mono-vs-color choice at print time. Why was this change made in cups-filters?

urf-supported is an AirPrint attribute and pwg-raster-document-type-supported is for IPP Everywhere/Wi-Fi Direct/Mopria support - seems like only having print-color-mode-supported will be the odd use case.

**Comment by zdohnal** [comment]
@michaelrsweet It was introduced due problems with at least two Canon printers (Canon ImageRunner Advance at the office and Canon MF645Cx from apple/cups#5713), where ```pwg-raster-document-type-supported``` attribute contained only ```srgb_8```, resulting into the problem created PPD didn't contain an option to change color mode into grayscale.
I was considering creating a 'template', meaning if ```pwg-raster-document-type-supported``` does provide colored option, but does not grayscale one, just add a grayscale one automatically, but we ended up with this solution.

**Review by michaelrsweet — CHANGES_REQUESTED** [review]
Given the description of the issue in the conversation, I think I would prefer having the code look for sgray_8, black_1, or black_8 when it sees srgb_8, and synthesize sgray_8 if there is no grayscale color space listed. The reason is that print-color-mode is effectively a rendering intent while pwg-raster-document-type-supported lists supported color spaces like AdobeRGB, sRGB, CMYK, etc. which are important for photo printing in particular.

**Comment by zdohnal** [comment]
The code is updated - I used ```ippContainsString()``` for checking for grayscale color models, and used ```ippGetCount()``` to check whether there is only one keyword. I thought it could speed up the execution if there is only one keyword.

Our Canon printer at the office got firmware update, which fixed the issue (got now ```srgb_8``` is listed twice in ```pwg-document-format-supported``` :( , but I would let it go and report it to Canon), so I'm not able to verify whether it works when only ```srgb_8``` is defined, but it doesn't generate a duplicate Grayscale entry at least and IMO there can be still printers in the wild with this error, so merging it makes sense.

There was an other reporter, which hit the issue with Canon, so I'll try to synch with him and see if we can verify the fix.

File: cups/ppd-cache.c [hunk 1]
```diff
@@ -3790,6 +3790,9 @@ _ppdCreateFromIPP(char   *buffer,	/* I - Filename buffer */
 	PRINTF_COLOROPTION("RGB", _("Color"), CUPS_CSPACE_SRGB, 8)
 
 	default_color = "RGB";
+
+	if (ippGetCount(attr) == 1 || !ippContainsString(attr, "sgray_8") && !ippContainsString(attr, "black_1") && !ippContainsString(attr, "black_8"))
```
Comment (michaelrsweet) [inline cups/ppd-cache.c:3794]: I think the count of 1 check is unnecessary given you are checking all of the other grayscale color spaces. But if we leave it in place I'd want to add parens around the ippContainsString checks (yes, it's fine for C order of operations but parens make it explicit visually...)

**Review by michaelrsweet — APPROVED** [review]
I think I'll drop the check for count == 1 but otherwise I like the new fix much better!

File: cups/ppd-cache.c (hunk 1 above)
Comment (zdohnal) [reply cups/ppd-cache.c:3794]: @michaelrsweet My intention with the count was the shortcut execution - meaning the code will not need to check for all strings if there is only 1 keyword - so ```ippGetCount()``` will return ```1``` and ```ippContainsString()``` calls will be skipped.

Did I misunderstand C condition's evaluation or do you think it is not worth it there?

Comment (michaelrsweet) [reply cups/ppd-cache.c:3794]: @zdohnal If the ippContainsString calls were expensive or this code was executed many times, then doing the initial shortcut evaluation would be worthwhile. But in this case it is a one-time set of calls so the extra optimization is unnecessary.

Comment (zdohnal) [reply cups/ppd-cache.c:3794]: Ok, got it! Thanks for the merging!

**Pull request merged**